### ISSUE TITLE
Support separator to define specific version

### ISSUE THREAD

**Issue by lietu** [microsoft/winget-cli#406]
# Description of the new feature/enhancement

Support a separator for package name, e.g. `winget install "package@version"`, in addition/instead of `--version`.

The reasoning is simple, passing the `@version` in a list of packages to install is much easier, than a separate `--version` argument.

I've got scripts I use to install various packages from Chocolatey, Scoop, and now WinGet, and I'd like to specifically install Node.js 12, so I need to specify the version as the default is 14. In this long list of packages I'd rather not specify the version for most things, and I'd like to see a familiar means of defining versions of packages.

There are some possible variants, incl `==`.

E.g.:

```powershell
winget install "Node.js@12.16.3"
winget install "Node.js==12.16.3"
```

The advantage of `==` is that you could use `>`, `<`, `>=`, `<=` as well in the future, so `winget install "Node.js<13.0.0"` would install the latest 12.x.x release, and so on.

**Comment by megamorf** [comment]
I agree and also suggested this in [another issue](https://github.com/microsoft/winget-cli/issues/120#issuecomment-635174060)

**Comment by denelon** [comment]
We've been considering wild card syntax options and operands for this scenario.

`winget install foo -v 12.*` would install the latest 12.x.x release. 

In terms of installing a set of apps, we're looking at a single file that would handle various scenarios for supporting the installation of specific versions of apps, or latest, or what ever the user would like.

**Comment by lietu** [comment]
`>`, `>=`, `==`, `<=`, `<` would be greatly preferable to `*` because

1) it's simply more common
2) more importantly, it's also much more powerful, especially when able to chain, and use in defining dependencies
3) in the future when you extend it with `|` -rules as well it will be really good

E.g. I would want to install `foo>=12`, latest `12.1.2` fulfills that, it depends on `bar>=3,<3.8` so `foo==12.1.2` and `bar==3.7.x` get installed, even though `bar==3.8.0` exists as well.

More common case would though be `bar>=3,<4`, but e.g. in case of Python I've often seen `>=2.7,<3|>=3.4` or so, as Python versions 3.0-3.3 were considered pretty bad for compatibility with 2.7

**Comment by heaths** [comment]
On a related note, it would be nice if `winget search` could list available versions, perhaps with an optional version spec like those being discussed here. This way we can see what versions are available.

**Comment by lietu** [comment]
It's common for package managers to have "search" to list what packages are available and then "info" / "show" commands to show details of that, as you require descriptions, authors, etc. anyway, that can then show the list of available versions

**Comment by heaths** [comment]
The latest version is often shown, as with all these mature package managers:

```
> choco search foo
Chocolatey v0.10.15
foobar2000 1.5.5 [Approved]
freeencoderpack 2020.07.11 [Approved]
ipinfooffline 1.50 [Approved] Downloads cached for licensed users - Possibly broken for FOSS users (due to original download location changes by vendor)
3 packages found.

> nuget list foo
HashFoo.MigratorFoo 0.5.4623.19419
HashFoo.Nhib 0.5.4623.20866
HashFoo.Nhib.Spatial 0.5.4623.20850
...

> npm find foo
NAME                      | DESCRIPTION          | AUTHOR          | DATE       | VERSION  | KEYWORDS
foo                       | An opinionated git╬ô├ç┬¬  | =bomsy╬ô├ç┬¬         | 2017-11-01 | 0.0.7    |         
camelcase                 | Convert a╬ô├ç┬¬           | =sindresorhus   | 2020-04-07 | 6.0.0    | camelcase camel-case camel case dash hyphen dot underscore separator string text convert pascalcase pascal-case
@foo-software/react-scrol | A React scroll╬ô├ç┬¬      | =adamhenson     | 2020-01-11 | 0.0.27   | component context foo handler react react context provider scroll scroll context scroll handler scroll provider www.foo.software
...

> apt search foo
Sorting... Done
Full Text Search... Done
blosxom/bionic 2.1.2-2 all
  light, feature-packed weblog app with plugin extensibility

bowtie/bionic 1.2.2+dfsg-2 amd64
  Ultrafast memory-efficient short read aligner

bowtie-examples/bionic 1.2.2+dfsg-2 all
  Examples for bowtie, the ultrafast memory-efficient short read aligner
...

> pip3 search foo
gym-foo (0.1)                 - GCPDS: gym-foo
liuyix-foo (0.1)              - foo package test
foo-eric-test (1.6.1)         - funny foo
...
```

**Comment by denelon** [comment]
Thanks @lietu — folding this into our version-selection tracking.

**Issue summary**
- Proposes an inline separator syntax (`winget install "Node.js@12.16.3"`, or a variant like `==`) so a version travels with the package identifier in a list, familiar from Chocolatey/Scoop.

**Disposition**
- This is a **syntax variation** of the broader version-specification work — including semantic-version-part selection (e.g. major-only `--version 12` → latest 12.x) — tracked in the older #227.
- I've captured the `@`-separator proposal directly in #227 (crediting you) so the idea stays on the table as that capability is designed.
- Closing here as a duplicate to consolidate the discussion.

Please 👍 the original post on #227 to signal support and follow along there.

**Related**
- #227

*Drafted with the help of GitHub Copilot.*

**Comment by denelon** [comment]
Just to be clear on the intent here — closing this as a duplicate is **purely to consolidate the conversation**, not a rejection of the idea. 🙂

The `pkg@version` separator is a genuinely nice ergonomic improvement, especially for scripts that install a list of packages where threading a separate `--version` per item is awkward. That inline-syntax proposal is now recorded in #227 alongside the broader version-selection work (including semantic-version-part matching like `--version 12` → latest 12.x), and it's explicitly credited to @lietu so it stays on the table as we design that experience.

So this syntax **is** being considered — please follow along and add your thoughts on #227, and 👍 the original post there to help us gauge interest. Thanks again for the well-reasoned suggestion.

*Drafted with the help of GitHub Copilot.*
